prime-router/src/main/kotlin/cli/ProcessHl7Commands.kt

```diff
@@ -0,0 +1,236 @@
+        if (comparisonFile.isBlank()) throw CliktError("File ${this.comparisonFile.absolutePath} is empty.")
+
+        val actionLogger = ActionLogger()
+        val starterMessages = HL7Reader(actionLogger).getMessages(starterFile)
+        val comparisonMessages = HL7Reader(actionLogger).getMessages(comparisonFile)
+
+        var counter = 0
+        starterMessages.forEach {
+            diffHl7(it, comparisonMessages[counter])
+            counter++
+        }
+    }
+
+    /**
+     * Used to keep track of the differences between segments. [segmentIndex] gives the place the difference is
+     * occurring at and [issue] holds the difference
+     */
+    data class Hl7Diff(val segmentIndex: String, val issue: String) {
+        override fun toString(): String {
+            return "Difference between messages at $segmentIndex: $issue"
+        }
+    }
+
+    /**
+     * Does the diffing of the [input] message to the [output] message. Results are echoed on the command line.
+     */
+    private fun diffHl7(input: Message, output: Message) {
+        val inputMap: MutableMap<String, Segment> = mutableMapOf()
+        val outputMap: MutableMap<String, Segment> = mutableMapOf()
+        val differences: MutableList<Hl7Diff> = mutableListOf()
+
+        /**
+         * Map the pieces of the structure to their index and name
+         */
+        fun indexStructure(structure: Structure, index: String, map: MutableMap<String, Segment>) {
+            when (structure) {
+                is Group -> {
+                    echo("is group")
+                    val childrenNames = structure.names.filter { cname -> structure.getAll(cname).isNotEmpty() }
+                    val children = childrenNames.map { structure.getAll(it) }
+                    children.forEach { childrenOfType ->
+                        childrenOfType.forEachIndexed { i, child ->
+                            echo("child: $child")
+                            indexStructure(child, "$index-${i + 1}", map)
+                        }
+                    }
+                }
+
+                is Segment -> {
+                    map["$index-${structure.name}"] = structure
+                }
+            }
+        }
+
+        val inputNames = input.names
+        inputNames.filter { name -> input.getAll(name).isNotEmpty() }.forEach { iname ->
+            val children = input.getAll(iname)
+            children.forEachIndexed { index, c ->
+                indexStructure(c, (index + 1).toString(), inputMap)
+            }
+        }
+        val outputNames = output.names
+        outputNames.filter { name -> output.getAll(name).isNotEmpty() }.forEach { iname ->
+            val children = output.getAll(iname)
+            children.forEachIndexed { index, c ->
+                indexStructure(c, (index + 1).toString(), outputMap)
+            }
+        }
+
+        fun compareHl7Type(
```

Comment:
This is what I originally started lifting out, but it has the echos in it that actually say what the values are that are different. I like having those, searching for them is even easier than counting the lines to the segment that has an issue, so I'd like to leave the echos in.

Reply:
So the echos are just the difference right?  One way you could keep those if rather than returning boolean, return `String?`.  If it's not null, everything matched, if it's there you can treat it as false and then echo it out.

Reply:
Yep, already got those changes in Friday and forgot to circle back to this comment